compiler/rustc_mir_transform/src/simplify_pow_of_two.rs

```diff
@@ -0,0 +1,212 @@
+//! A pass that checks for and simplifies calls to `pow` where the receiver is a power of
+//! two. This can be done with `<<` instead.
+
+use crate::MirPass;
+use rustc_const_eval::interpret::{ConstValue, Scalar};
+use rustc_middle::mir::patch::MirPatch;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt, UintTy};
+use rustc_span::sym;
+use rustc_target::abi::FieldIdx;
+
+pub struct SimplifyPowOfTwo;
+
+impl<'tcx> MirPass<'tcx> for SimplifyPowOfTwo {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut patch = MirPatch::new(body);
+
+        for (i, bb) in body.basic_blocks.iter_enumerated() {
+            let term = bb.terminator();
+            let source_info = term.source_info;
+            let span = source_info.span;
+
+            if let TerminatorKind::Call {
+                    func,
+                    args,
+                    destination,
+                    target: Some(target),
+                    call_source: CallSource::Normal,
+                    ..
+                } = &term.kind
+                && let Some(def_id) = func.const_fn_def().map(|def| def.0)
+                && let def_path = tcx.def_path(def_id)
+                && tcx.crate_name(def_path.krate) == sym::core
+                && let [recv, exp] = args.as_slice()
+                && let Some(recv_const) = recv.constant()
+                && let ConstantKind::Val(
+                    ConstValue::Scalar(Scalar::Int(recv_int)),
+                    recv_ty,
+                ) = recv_const.literal
+                && recv_ty.is_integral()
+                && tcx.item_name(def_id) == sym::pow
+                && let Ok(recv_val) = match recv_ty.kind() {
+                    ty::Int(_) => {
+                        let result = recv_int.try_to_int(recv_int.size()).unwrap_or(-1).max(0);
+                        if result > 0 {
+                            Ok(result as u128)
+                        } else {
+                            continue;
+                        }
+                    },
+                    ty::Uint(_) => recv_int.try_to_uint(recv_int.size()),
+                    _ => continue,
+                }
+                && let power_used = f32::log2(recv_val as f32)
+                // Precision loss means it's not a power of two
+                && power_used == (power_used as u32) as f32
```

Comment:
Why use floats instead of ilog2?

Reply:
There's unfortunately no distinction between whether it's power of two there. `i32::ilog2(4)` and `i32::log2(5)` both return 5.